### DISCUSSION TITLE
Bazel Examples

### DISCUSSION THREAD

**Discussion by radvani13** [bazelbuild/bazel#18709] [Polls]
Hi all,

We heard from our CSAT survey, that lack of good examples and tutorials is big pain point for most of Bazel users. In an effort to improve the examples we would like to learn which of these are more important to you? Please answer the poll so we can target the top examples. If we are missing examples - please reach on this post and I will add

**Comment by sluongng** [comment]
Given https://github.com/bazelbuild/bazel/discussions/18626 and https://github.com/bazelbuild/bazel/discussions/18546 and https://github.com/bazelbuild/bazel/discussions/18332, you have my vote on a hermetic cc toolchain example.

**Reply by radvani13** [reply]
@sluongng  thank you also updated the poll - since Its still just fresh

**Comment by aschleck** [comment]
Thank you Radhika for starting this! The poll only allows single choice so I voted for "bzlmod hermetic cc" but all 3 of the bzlmod items would be extremely helpful for me as a reference. In terms of structuring examples, I feel like having a complete working example folder is the most important missing piece. Right now I still am not certain if a fully migrated bzlmod project should have a "WORKSPACE" file or not, nor do I know if that WORKSPACE file should contain `workspace(...)` or not.

Others I can think of:
* Bzlmod: hermetic Java example (I currently pull in both @io_tweag_rules_nixpkgs and @remote_java_tools in my project, not sure why I need @remote_java_tools at all) including bazelrc flags to make Bazel use it in all cases
* Bzlmod: how to patch a bzlmod dependency (I believe there's already an example of this somewhere but I can no longer find it)
* Bzlmod: how to patch the registry definition of a bzlmod dependency (not sure if this is possible, but I get warnings due to protobuf in the BCR specifying an older Maven Guava artifact than my project specifies in my MODULE.bazel and it'd be nice to make it go away)

Regarding "How to build a java, protobuf and python app with Bazel?", if you were to do this I would personally think protobuf+gRPC would be more valuable than just protobuf alone. Last time I looked there was drama regarding py_proto_library and I couldn't figure out what the accepted solution was (I gave up and wrote my own Bazel actions to do it instead.)

**Reply by meteorcloudy** [reply]
> but I get warnings due to protobuf in the BCR specifying an older Maven Guava artifact than my project specifies in my MODULE.bazel and it'd be nice to make it go away

This warning is likely coming from rules_jvm_external, can you file an issue there to for reducing the warning messages?

**Reply by radvani13** [reply]
@shs96c i think :) for rules_jvm_external

**Comment by lbhdc** [comment]
I wish this poll were multi-choice. I want so many of these.
These are what I would select :grin:
- hermetic cc-toolchain that works with sanitizers would be huge. I see lots of questions about it here, and I really struggle with the docs. Having a concrete example would be a huge help. Sanitizers are a huge productivity boost when writing c++ and help avoid a lot of the unsafe parts of the language. An example that shows how to make a hermetic toolchain that works with sanitizers (and native cc_test rules) would be invaluable.

- bzlmod: Use local version of repository. I am not sure if this means a self hosted bcr, or in-tree modules (a la local_workspace). I want both. I want to be able to self all my third part deps. Currently this is super easy, you just fork on the VCS provider you like and update your links. I also want to be able to carve my monorepos up into modules. There are some deps that take crazy long to compile (looking at you grpc) that I think (hope) I could reduce the need to compile by splitting things up.
- bzlmod: using repos that aren't in bcr. Over the years I have accumulated a bunch of build files for random repos that assimilate them into bazel (I can share some if examples would help). Being able to use them as modules translate that into the new world will be very helpful.

Others I would love to see:
- Patching modules. This is something I find essential to make bazel work, particularly with cc deps, and rule sets. It would be really helpful to see how to patch a module (including things like module deps).

- Setting up generic toolchains and things that look like toolchains. Things like the openpolicyagent tool chain to test rego files as well as homegrown toolchains.
